annotation-processor/src/main/kotlin/pl/touk/krush/source/MappingsGenerator.kt

```diff
@@ -127,37 +209,185 @@ abstract class MappingsGenerator : SourceGenerator {
+                        val targetColumnName = "${entity.name.asVariable()}Target${property.valName.capitalize()}"
+                        "\t\t$name = this[$relationTableName.$targetColumnName]"
+                    }
+                    val nullCheck = id.properties.joinToString(" && ") { property ->
+                        val targetColumnName = "${entity.name.asVariable()}Target${property.valName.capitalize()}"
+                        "this.getOrNull($relationTableName.$targetColumnName) == null"
+                    }
+                    "if($nullCheck) null else ${id.qualifiedName}(\n$embeddableIdMapping\n\t)"
+
+                } else {
+                    val targetColumnName = "${entity.name.asVariable()}Target${entity.id.name.toString().capitalize()}"
+                    "this.getOrNull($relationTableName.$targetColumnName)"
+                }
+
+                val selfReferenceMapName = "${entity.name.asVariable()}SelfReferenceRequests"
+
+                func.apply {
+                    // Allowing a null id here allows users to not include a join with the other table if they don't
+                    // need the relation-lists to be populated
+                    addStatement("val other${entity.name}Id = $idReadingBlock")
+                    addStatement("if(other${entity.name}Id != null) {")
+                    addStatement("\tif(selfReferenceRequests[\"${entity.name}\"] == null) selfReferenceRequests[\"${entity.name}\"] = mutableMapOf()")
+                    addStatement("\tval $selfReferenceMapName = selfReferenceRequests[\"${entity.name}\"]!!")
+                    addStatement("\tif($selfReferenceMapName[other${entity.name}Id] == null) $selfReferenceMapName[other${entity.name}Id] = mutableSetOf()")
+                    addStatement("\t$selfReferenceMapName[other${entity.name}Id]!!.add($entityParamName.${id.name}!!)")
+                    addStatement("}")
+                }
+            }
+        }
+
+        return func.build()
+    }
+
+    private fun buildToEntityMapFunc(entityType: TypeElement, entity: EntityDefinition, graph: EntityGraph): FunSpec {
         val rootKey = entity.id?.asUnderlyingTypeName() ?: throw MissingIdException(entity)
 
-        val rootVal = entity.name.asVariable()
         val func = FunSpec.builder("to${entity.name}Map")
                 .receiver(Iterable::class.parameterizedBy(ResultRow::class))
-                .returns(ClassName("kotlin.collections", "MutableMap").parameterizedBy(rootKey, entityType.toImmutableKmClass().toClassName()))
+                .returns(
+                    ClassName("kotlin.collections", "Map")
+                        .parameterizedBy(rootKey, entityType.toImmutableKmClass().toClassName())
+                )
+        
+        val currentEntityValName = "current${entity.name.asObject()}"
 
-        val rootIdName = entity.id.name.asVariable()
-        val rootValId = "${rootVal}Id"
+        func.apply {
```

Comment:
I like the general idea of EntityStore and thought having some wrapper class holding current ResultRow, entity store and selfReferencesRequest:
```kotlin
data class RowWrapper(
  val row: ResultRow,
  val entityStore: MutableMap<String, MutableMap<Any, Any>>,
  val selfReferenceRequests: MutableMap<String, MutableMap<Any, MutableSet<Any>>>
)
```
We can now move `toXXX` and `addSubEntitiesToXXX` methods to this wrapper class to avoid passing stores as arguments.

Reply:
And maybe we can have `KClass` as key in `entityStore`?

Reply:
I think it's a good idea to "summarize" these maps in a data class. I think it would be better to make the row a mutable property though, because otherwise the object would have to be recreated for every row that is being iterated over while the maps stay the same anyways.

One potential issue that comes to mind is that making `toEntity` a method of `RowWrapper` would essentially make it an internal method (at least I don't expect users of the API to use it or the entity maps). However, it does feel like a method that may be useful to end users, too if they want to convert a single result row to an entity for some reason. On the other hand, `toEntity` doesn't parse the non-O2O-relations anyways, so it might not be that useful standalone after all. I'm not sure about this.

By moving methods to the wrapper class, do you mean making them part of the wrapper class itself, or making them extension functions of it? The former would raise the question of where to put that class, since there is currently one mapping-file per package with entities, but relations can be cross-package.

> And maybe we can have KClass as key in entityStore?

That is a good idea, it makes the purpose of the map a little less opaque (having to use `Any` is already unfortunate). The same thing could be done for `selfReferenceRequests`, since that uses class names as a key as well.

Reply:
- regarding `toEntity` I agree it should be public, so let's leave `entityStore` as argument
 - I thought about extension methods in `RowWrapper` - in the next step we could try to generate separate `entityStore`s for each class, as we know exact types we're going to load in each query - do you think it would be possible?

Reply:
If we use a separate `entityStore` for every type, we'd have to pass all of them as arguments in `toEntity`, if `toEntity` isn't an extension method of `RowWrapper`.
Maybe a middle way will do the trick: How about we make the current `toEntity` an extension method of `RowWrapper`, but add a separate version for `ResultRow` that uses the `RowWrapper`? That way we'd get the advantage of both variants.

Reply:
I wonder if you could turn on "Allow edits from maintainers" on your fork as in [this article](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) so we could work together on this code?

Reply:
Unless I'm missing something, that is already turned on in the PR view.

Reply:
Ok, I thought I can push directly via PR, but had to fetch also your fork - and now it works! I fixed the tests and added one more test case.

Reply:
I just started implementing the `RowWrapper` with a map for each entity. However, for the example project, that resulted in a data class with ~100 maps, some of which have conflicting names/imports. For the example project, the solution to that would be generating a RowWrapper-class per package/graph. However, I don't think we can generally assume that all entity classes share a package.

Reply:
Maybe we can have shared `RowWrapper` in `krush-runtime` module?

Reply:
I pushed also cleaned up template of generated code to `TaskTest` https://github.com/TouK/krush/pull/48/commits/350dc401da0f9f3c6a241be683db47854480bff8#diff-c9c83d541492af36db7a4bb94e3341f7be8c368792f5ff6b907f713a46b57613R53

Reply:
> Maybe we can have shared `RowWrapper` in `krush-runtime` module?

Sure, we'd just have to use the `KClass` approach instead of generating maps. I would prefer using properly typed maps, if at all possible.